File: Data Structures Intro/Binary Search Tree/Java/binary-search-tree-task-java/src/main/java/com/epam/bsp/binary/tree/TreeNode.java
package com.epam.bsp.binary.tree;

public class TreeNode<E extends Comparable<E>> {

    /**
     * Data part
     */
    private E element;
    /**
     * Links to the left and right nodes
     */
    private TreeNode<E> left;
    private TreeNode<E> right;

    public TreeNode(E element, TreeNode<E> left, TreeNode<E> right) {
        this.element = element;
        this.left = left;
        this.right = right;
    }

    public TreeNode(E element) {
        this(element, null, null);
    }

    public TreeNode(E element, TreeNode<E> left) {
        this(element, left, null);
    }

    public E getElement() {
        return element;
    }

    public TreeNode<E> getLeft() {
        return left;
    }

    public TreeNode<E> getRight() {
        return right;
    }

    public void setElement(E element) {
        this.element = element;
    }

    public void setLeft(TreeNode<E> left) {
        this.left = left;
    }

    public void setRight(TreeNode<E> right) {
        this.right = right;
    }

    /**
     * Checks whether the tree is height-balanced.
     * A tree is balanced if the height difference of left and right subtrees
     * of every node is not more than 1.
     */
    public boolean isBalanced() {
        return checkBalance(this) != -1;
    }

    private int checkBalance(TreeNode<E> node) {
        if (node == null) return 0;
        int leftHeight = checkBalance(node.left);
        if (leftHeight == -1) return -1;
        int rightHeight = checkBalance(node.right);
        if (rightHeight == -1) return -1;

        if (Math.abs(leftHeight - rightHeight) > 1) return -1;
        return Math.max(leftHeight, rightHeight) + 1;
    }

    /**
     * Checks whether the tree is a valid Binary Search Tree.
     */
    public boolean isBinarySearchTree() {
        return isBST(this, null, null);
    }

    private boolean isBST(TreeNode<E> node, E min, E max) {
        if (node == null) return true;

        E val = node.getElement();
        if ((min != null && val.compareTo(min) <= 0) ||
            (max != null && val.compareTo(max) >= 0)) {
            return false;
        }

        return isBST(node.left, min, val) && isBST(node.right, val, max);
    }

    /**
     * Searches for an element in the BST and returns the node containing it.
     */
    public TreeNode<E> searchInBst(E element) {
        if (this == null) return null;
        int cmp = element.compareTo(this.element);

        if (cmp == 0) return this;
        else if (cmp < 0 && left != null) return left.searchInBst(element);
        else if (cmp > 0 && right != null) return right.searchInBst(element);
        return null;
    }

    /**
     * Inserts a new element into the BST and returns the root of the tree.
     */
    public TreeNode<E> insertInBst(E element) {
        if (element.compareTo(this.element) < 0) {
            if (this.left == null) {
                this.left = new TreeNode<>(element);
            } else {
                this.left.insertInBst(element);
            }
        } else if (element.compareTo(this.element) > 0) {
            if (this.right == null) {
                this.right = new TreeNode<>(element);
            } else {
                this.right.insertInBst(element);
            }
        }
        return this;
    }
}

File: Data Structures Intro/Binary Search Tree/C#/binary-search-tree-task-net/BinarySearchTree/TreeNode.cs
using System;

namespace BinarySearchTreeTask;

/// <summary>
/// Represents a node in a binary tree with generic type support
/// </summary>
/// <typeparam name="T">The type of the element stored in the node, must implement IComparable</typeparam>
public class TreeNode<T> : IBinarySearchTree<T> where T : IComparable<T>
{
    private T element;
    private TreeNode<T>? left;
    private TreeNode<T>? right;

    public TreeNode(T element, TreeNode<T>? left, TreeNode<T>? right)
    {
        Element = element ?? throw new ArgumentNullException(nameof(element));
        Left = left;
        Right = right;
    }

    public TreeNode(T element) : this(element, null, null) { }

    public TreeNode(T element, TreeNode<T>? left) : this(element, left, null) { }

    public T Element
    {
        get => element;
        private set => element = value ?? throw new ArgumentNullException(nameof(value));
    }

    public TreeNode<T>? Left
    {
        get => left;
        private set => left = value;
    }

    public TreeNode<T>? Right
    {
        get => right;
        private set => right = value;
    }

    public bool IsBalanced()
    {
        return CheckBalance(this).isBalanced;
    }

    private (bool isBalanced, int height) CheckBalance(TreeNode<T>? node)
    {
        if (node == null)
            return (true, 0);

        var leftResult = CheckBalance(node.Left);
        if (!leftResult.isBalanced)
            return (false, 0);

        var rightResult = CheckBalance(node.Right);
        if (!rightResult.isBalanced)
            return (false, 0);

        bool balanced = Math.Abs(leftResult.height - rightResult.height) <= 1;
        int height = Math.Max(leftResult.height, rightResult.height) + 1;
        return (balanced, height);
    }

    public bool IsBinarySearchTree()
    {
        return ValidateBst(this, default, default, false, false);
    }

    private bool ValidateBst(TreeNode<T>? node, T? min, T? max, bool hasMin, bool hasMax)
    {
        if (node == null)
            return true;

        // node.Element must be > min (strictly greater)
        if (hasMin && node.Element.CompareTo(min!) < 0)
            return false;

        // node.Element must be <= max (allow duplicates on right)
        if (hasMax && node.Element.CompareTo(max!) > 0)
            return false;

        return ValidateBst(node.Left, min, node.Element, hasMin, true)
            && ValidateBst(node.Right, node.Element, max, true, hasMax);
    }

    public TreeNode<T>? Search(T element)
    {
        if (element == null)
            throw new ArgumentNullException(nameof(element));

        TreeNode<T>? current = this;
        while (current != null)
        {
            int cmp = element.CompareTo(current.Element);
            if (cmp == 0)
                return current;
            else if (cmp < 0)
                current = current.Left;
            else
                current = current.Right;
        }
        return null;
    }

    public TreeNode<T>? SearchInBst(T element) => Search(element);

    public TreeNode<T> Insert(T element)
    {
        if (element == null)
            throw new ArgumentNullException(nameof(element));

        InsertRecursive(this, element);
        return this;
    }

    private TreeNode<T> InsertRecursive(TreeNode<T> node, T element)
    {
        if (element == null)
            throw new ArgumentNullException(nameof(element));

        int cmp = element.CompareTo(node.Element);

        if (cmp < 0)
        {
            node.Left = node.Left == null ? new TreeNode<T>(element) : InsertRecursive(node.Left, element);
        }
        else if (cmp > 0)
        {
            node.Right = node.Right == null ? new TreeNode<T>(element) : InsertRecursive(node.Right, element);
        }
        else
        {
            return node;
        }

        return node;
    }

    public TreeNode<T> InsertInBst(T element) => Insert(element);
}